File: nchito-ios/Nchito/Models/WireFormat.swift
import Foundation

// Translation between the app's domain enums and the values Postgres actually
// stores, defined in supabase/migrations/.
//
// This file exists because the enums carry DISPLAY text as their rawValue —
// `GigCategory.delivery` is "Delivery & Errands" — while the database enum is
// 'delivery'. Encoding a model straight to JSON would therefore send text the
// database rejects, or worse, quietly write nonsense. Every enum that crosses
// the wire needs an explicit mapping here, and `WireFormatTests` checks the
// round trip so a renamed label can't silently break writes.

protocol WireRepresentable: CaseIterable, Hashable {
    /// Exactly the value the Postgres enum stores.
    var wireValue: String { get }
}

extension WireRepresentable {
    static func from(wire: String) -> Self? {
        allCases.first { $0.wireValue == wire }
    }
}

extension GigCategory: WireRepresentable {
    var wireValue: String {
        switch self {
        case .delivery:     return "delivery"
        case .homeServices: return "home_services"
        case .tutoring:     return "tutoring"
        case .digital:      return "digital"
        case .events:       return "events"
        case .farm:         return "farm"
        case .beauty:       return "beauty"
        case .repairs:      return "repairs"
        }
    }
}

extension GigStatus: WireRepresentable {
    static var allCases: [GigStatus] { [.open, .assigned, .completed, .paid] }

    var wireValue: String {
        switch self {
        case .open:      return "open"
        case .assigned:  return "assigned"
        case .completed: return "completed"
        case .paid:      return "paid"
        }
    }
}

extension MicroTaskKind: WireRepresentable {
    var wireValue: String {
        switch self {
        case .survey:      return "survey"
        case .appTest:     return "app_test"
        case .dataLabel:   return "data_label"
        case .social:      return "social"
        case .mysteryShop: return "mystery_shop"
        }
    }
}

extension TransactionKind: WireRepresentable {
    static var allCases: [TransactionKind] {
        [.gigPayout, .taskReward, .referralBonus, .cashOut, .boostPurchase,
         .wageAdvance, .advanceRepayment, .advanceRecovery, .escrowIn, .escrowRefund]
    }

    var wireValue: String {
        switch self {
        case .gigPayout:        return "gig_payout"
        case .taskReward:       return "task_reward"
        case .referralBonus:    return "referral_bonus"
        case .cashOut:          return "cash_out"
        case .boostPurchase:    return "boost_purchase"
        case .wageAdvance:      return "wage_advance"
        case .advanceRepayment: return "advance_repayment"
        case .advanceRecovery:  return "advance_recovery"
        case .escrowIn:         return "escrow_in"
        case .escrowRefund:     return "escrow_refund"
        case .other:            return "other"
        }
    }
}

extension TransactionKind {
    /// Never fails: a kind this build doesn't recognise becomes `.other`, so one
    /// unfamiliar row can't take down a user's whole transaction list.
    static func lenient(wire: String) -> TransactionKind {
        from(wire: wire) ?? .other
    }
}

extension VerificationLevel: WireRepresentable {
    static var allCases: [VerificationLevel] { [.unverified, .phoneVerified, .nrcVerified] }

    var wireValue: String {
        switch self {
        case .unverified:    return "unverified"
        case .phoneVerified: return "phone_verified"
        case .nrcVerified:   return "nrc_verified"
        }
    }
}

extension ProofPhoto.Kind: WireRepresentable {
    var wireValue: String { rawValue }   // already 'before' / 'after'
}

extension WageAdvance.Status: WireRepresentable {
    static var allCases: [WageAdvance.Status] {
        [.outstanding, .settled, .recovering, .writtenOff]
    }
    var wireValue: String { rawValue }   // already matches advance_status
}

// MARK: - Dates

enum WireDate {
    /// Postgres returns timestamptz with a variable number of fractional digits,
    /// which `.iso8601` alone rejects — so try with fractional seconds first.
    static let withFraction: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    static let plain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    static func parse(_ string: String) -> Date? {
        withFraction.date(from: string) ?? plain.date(from: string)
    }

    static func format(_ date: Date) -> String {
        withFraction.string(from: date)
    }
}

extension JSONDecoder {
    /// Decoder configured for PostgREST responses.
    static var nchito: JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let raw = try decoder.singleValueContainer().decode(String.self)
            guard let date = WireDate.parse(raw) else {
                throw DecodingError.dataCorruptedError(
                    in: try decoder.singleValueContainer(),
                    debugDescription: "Unrecognised timestamp: \(raw)")
            }
            return date
        }
        return decoder
    }
}

extension JSONEncoder {
    static var nchito: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .custom { date, encoder in
            var container = encoder.singleValueContainer()
            try container.encode(WireDate.format(date))
        }
        return encoder
    }
}

File: nchito-android/app/src/main/java/com/owenphiri/nchito/data/WireFormat.kt
package com.owenphiri.nchito.data

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Translation between the app's domain enums and the values Postgres actually
 * stores, defined in nchito-ios/supabase/migrations/.
 *
 * This exists because the enums carry DISPLAY text — `GigCategory.DELIVERY.label`
 * is "Delivery & Errands" — while the database enum is 'delivery'. Sending the
 * label would be rejected, or worse, quietly write nonsense. Every enum that
 * crosses the wire is mapped explicitly here.
 */

// GigCategory carries its own wire value (generated in ServiceCatalog.kt).
// These two aliases keep it addressable through the same names as every
// other enum here, so call sites do not have to know which is which.
val GigCategory.wireValue: String get() = wire

fun gigCategoryFromWire(wire: String): GigCategory? = GigCategory.fromWire(wire)

val GigStatus.wireValue: String
    get() = when (this) {
        GigStatus.OPEN -> "open"
        GigStatus.ASSIGNED -> "assigned"
        GigStatus.COMPLETED -> "completed"
        GigStatus.PAID -> "paid"
    }

fun gigStatusFromWire(wire: String): GigStatus? =
    GigStatus.entries.firstOrNull { it.wireValue == wire }

val MicroTaskKind.wireValue: String
    get() = when (this) {
        MicroTaskKind.SURVEY -> "survey"
        MicroTaskKind.APP_TEST -> "app_test"
        MicroTaskKind.DATA_LABEL -> "data_label"
        MicroTaskKind.SOCIAL -> "social"
        MicroTaskKind.MYSTERY_SHOP -> "mystery_shop"
    }

fun microTaskKindFromWire(wire: String): MicroTaskKind? =
    MicroTaskKind.entries.firstOrNull { it.wireValue == wire }

val TxKind.wireValue: String
    get() = when (this) {
        TxKind.GIG_PAYOUT -> "gig_payout"
        TxKind.TASK_REWARD -> "task_reward"
        TxKind.REFERRAL_BONUS -> "referral_bonus"
        TxKind.CASH_OUT -> "cash_out"
        TxKind.WAGE_ADVANCE -> "wage_advance"
        TxKind.ADVANCE_REPAYMENT -> "advance_repayment"
        TxKind.ADVANCE_RECOVERY -> "advance_recovery"
        TxKind.ESCROW_IN -> "escrow_in"
        TxKind.ESCROW_REFUND -> "escrow_refund"
        TxKind.OTHER -> "other"
    }

/**
 * Never fails: a kind this build doesn't recognise becomes OTHER, so one
 * unfamiliar row can't take down a user's whole transaction list. Mobile
 * versions live in the wild for a long time after a migration adds a value.
 */
fun txKindFromWire(wire: String): TxKind =
    TxKind.entries.firstOrNull { it.wireValue == wire } ?: TxKind.OTHER

val AdvanceStatus.wireValue: String
    get() = when (this) {
        AdvanceStatus.OUTSTANDING -> "outstanding"
        AdvanceStatus.SETTLED -> "settled"
        AdvanceStatus.RECOVERING -> "recovering"
        AdvanceStatus.WRITTEN_OFF -> "written_off"
    }

fun advanceStatusFromWire(wire: String): AdvanceStatus =
    AdvanceStatus.entries.firstOrNull { it.wireValue == wire } ?: AdvanceStatus.OUTSTANDING

val ProofKind.wireValue: String
    get() = if (this == ProofKind.BEFORE) "before" else "after"

fun proofKindFromWire(wire: String): ProofKind =
    if (wire == "after") ProofKind.AFTER else ProofKind.BEFORE

/**
 * Postgres returns timestamptz with a variable number of fractional digits, so
 * parsing tries the fractional form first and falls back.
 */
object WireDate {
    private fun formatter(pattern: String) =
        SimpleDateFormat(pattern, Locale.US).apply { timeZone = TimeZone.getTimeZone("UTC") }

    private val patterns = listOf(
        "yyyy-MM-dd'T'HH:mm:ss.SSSXXX",
        "yyyy-MM-dd'T'HH:mm:ssXXX",
        "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
        "yyyy-MM-dd'T'HH:mm:ss'Z'",
    )

    fun parse(value: String): Date? {
        for (p in patterns) {
            runCatching { return formatter(p).parse(value) }
        }
        return null
    }

    fun format(date: Date): String = formatter("yyyy-MM-dd'T'HH:mm:ss.SSSXXX").format(date)
}
